lib/global-styles.php

```diff
@@ -164,15 +164,90 @@ function gutenberg_experimental_global_styles_get_theme() {
 }
 
 /**
- * Takes a Global Styles tree and returns a CSS rule
- * that contains the corresponding CSS custom properties.
+ * Given a base line height and a multiplier, generates a new line height.
+ *
+ * @param integer $line_height Base line height.
+ * @param integer $multiplier Multiplier.
+ * @return float New line height.
+ */
+function gutenberg_experimental_global_styles_generate_line_height( $line_height, $multiplier ) {
+	return round( $line_height * $multiplier, 2 );
+}
+
+/**
+ * Given a base font weight and a multiplier, generates a new font weight.
+ *
+ * @param integer $font_weight Base font weight.
+ * @param integer $multiplier Multiplier.
+ * @return integer New font weight.
+ */
+function gutenberg_experimental_global_styles_generate_font_weight( $font_weight, $multiplier ) {
+	return intval( $font_weight * $multiplier );
+}
+
+/**
+ * Given a font size, scale, and multiplier,
+ * generates a new font size.
+ *
+ * @param string  $font_base Base font size in pixels.
+ * @param integer $font_scale Base scale.
+ * @param integer $multiplier Multiplier.
+ * @return float Font size in pixels.
+ */
+function gutenberg_experimental_global_styles_generate_font_size( $font_base, $font_scale, $multiplier ) {
+	$font_base_unitless = intval( str_replace( 'px', '', $font_base ) );
+	return round( pow( $font_scale, $multiplier ) * $font_base_unitless, 2 ) . 'px';
+}
+
+/**
+ * Given a global styles tree, generates the value of some
+ * design tokens.
  *
  * @param array $global_styles Global Styles tree.
+ * @return array Global Styles tree with the new values.
+ */
+function gutenberg_experimental_global_styles_generate_styles( $global_styles ) {
+	$line_height = $global_styles['typography']['line-height'];
+	$font_size   = $global_styles['typography']['font-size'];
+	$font_scale  = $global_styles['typography']['font-scale'];
+	$font_weight = $global_styles['typography']['font-weight'];
+
+	$tree = array_replace_recursive(
+		$global_styles,
+		array(
+			'typography' => array(
+				'line-height-heading' => gutenberg_experimental_global_styles_generate_line_height( $line_height, 0.8 ),
```

Comment:
Can't we use CSS calc https://developer.mozilla.org/en-US/docs/Web/CSS/calc on the CSS end to generate these values?

E.g: core would set variable for font-size-heading-3 equal to calc depending on another variable. Themes would be able to overwrite that.

Reply:
See https://github.com/WordPress/gutenberg/issues/21030

Reply:
See https://github.com/WordPress/gutenberg/pull/20530#discussion_r391767806